src/Compilers/Core/MSBuildTask/MapSourceRoots.cs

```diff
@@ -0,0 +1,166 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Tasks;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.CodeAnalysis.BuildTasks
+{
+    /// <summary>
+    /// Given a list of SourceRoot items produces a list of the same items with added <c>MappedPath</c> metadata that
+    /// contains calculated deterministic source path for each SourceRoot.
+    /// </summary>
+    /// <remarks>
+    /// Does not perform any path validation.
+    /// </remarks>
+    public sealed class MapSourceRoots : Task
+    {
+        public MapSourceRoots()
+        {
+            TaskResources = ErrorString.ResourceManager;
+        }
+
+        /// <summary>
+        /// SourceRoot items with the following optional well-known metadata:
+        /// <list type="bullet">
+        ///   <term>SourceControl</term><description>Indicates name of the source control system the source root is tracked by (e.g. Git, TFVC, etc.), if any.</description>
+        ///   <term>NestedRoot</term><description>If a value is specified the source root is nested (e.g. git submodule). The value is a path to this root relative to the containing root.</description>
+        ///   <term>ContainingRoot</term><description>Identifies another source root item that this source root is nested under.</description>
+        /// </list>
+        /// </summary>
+        [Required]
+        public ITaskItem[] SourceRoots { get; set; }
+
+        /// <summary>
+        /// True if the mapped paths should be deterministic.
+        /// </summary>
+        public bool Deterministic { get; set; }
+
+        [Output]
+        public ITaskItem[] MappedSourceRoots { get; private set; }
+
+        private static class Names
+        {
+            public const string SourceRoot = nameof(SourceRoot);
+
+            // Names of well-known SourceRoot metadata items:
+            public const string SourceControl = nameof(SourceControl);
+            public const string NestedRoot = nameof(NestedRoot);
+            public const string ContainingRoot = nameof(ContainingRoot);
+            public const string MappedPath = nameof(MappedPath);
+        }
+
+        public override bool Execute()
+        {
+            var topLevelMappedPaths = new Dictionary<string, string>();
+            int i = 0;
+
+            void SetTopLevelMappedPaths(bool sourceControl)
```

Comment:
Nit: local functions should be camelCased. If it's ok, please also move to the end of the method.

Reply:
I have actually refactored it a bit here.

Reply:
Will change the casing.

Reply:
I'm looking at iteration 3. Should I wait until you push something more?

---
In reply to: [176874575](https://github.com/dotnet/roslyn/pull/25325#discussion_r176874575) [](ancestors = 176874575)